torch/cuda/amp/grad_scaler.py

```diff
@@ -329,7 +331,30 @@ def step(self, optimizer, *args, **kwargs):
             # The contract with custom optimizers is that their step() should accept an additional,
             # optional grad_scaler kwarg.  We append self to the kwargs so the custom optimizer has full information:
             # it can query its own state, invoke unscale_ on itself, etc
-            retval = optimizer.step(*args, **dict(kwargs, grad_scaler=self))
+            # The contract above is being deprecated to avoid introducing `grad_scaler: GradScaler` argument
+            # to `Optimizer.step`. The new behavior is going to add two Tensor attributes of `grad_scale`
+            # and `found_inf` to the passed optimizer so that the optimizer can utilize those
+            # to skip the parameter updates or unscale gradients before updating parameters in
+            # the fused kernel, e.g. `FusedAdamMathFunctor`.
+            kwargs_ = kwargs
+            if "grad_scaler" in inspect.signature(optimizer.step).parameters:
+                warnings.warn(
+                    "GradScaler is going to stop passing itself as a keyword argument to the passed "
+                    "optimizer. In the near future GradScaler registers `grad_scale: Tensor` and "
+                    "`found_inf: Tensor` to the passed optimizer and let the optimizer use them directly.",
+                    FutureWarning)
+                kwargs_.update({"grad_scaler": self})
+            else:
+                scaler = self._get_scale_async()
+                found_inf = cast(
+                    torch.Tensor,
+                    sum([
+                        t.to(scaler.device, non_blocking=True) for t in self._check_inf_per_device(optimizer).values()
+                    ])
+                )
+                optimizer.grad_scale = scaler
```

Comment:
This doesn't have to change but there might be an existing grad_scale and found_inf fields. Do we want to re-use them in the future or we're happy to creating new Tensors for these at every step?

If the second, should be remove them from the optimizer state just after the step to reduce memory?